File: app/src/main/java/n7/ad2/ui/streams/StreamsPagedListAdapter.java
package n7.ad2.ui.streams;

import androidx.paging.PagedListAdapter;
import androidx.databinding.DataBindingUtil;
import android.os.Handler;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.robinhood.ticker.TickerUtils;

import java.util.Random;

import n7.ad2.R;
import n7.ad2.data.source.remote.model.Stream;
import n7.ad2.databinding.ItemListStreamBinding;

public class StreamsPagedListAdapter extends PagedListAdapter<Stream, StreamsPagedListAdapter.ViewHolder> {

    private static final DiffUtil.ItemCallback<Stream> DIFF_CALLBACK = new DiffUtil.ItemCallback<Stream>() {
        @Override
        public boolean areItemsTheSame(@NonNull Stream streams, @NonNull Stream t1) {
            return true;
        }

        @Override
        public boolean areContentsTheSame(@NonNull Stream streams, @NonNull Stream t1) {
            return false;
        }
    };
    private final int[] view = {-2, -1, 0, 1, 2, 3};
    private final long[] duration = {2000, 3000, 4000, 3500, 2500};
    private LayoutInflater inflater;

    StreamsPagedListAdapter() {
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (inflater == null) inflater = LayoutInflater.from(viewGroup.getContext());

        ItemListStreamBinding binding = DataBindingUtil.inflate(inflater, R.layout.item_list_stream, viewGroup, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int i) {

    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ItemListStreamBinding binding;
        Handler handler;

        ViewHolder(@NonNull final ItemListStreamBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            binding.tvItemListStreamViewers.setCharacterList(TickerUtils.getDefaultNumberList());
            handler = new Handler();
            handler.post(new Runnable() {
                @Override
                public void run() {
                    try {
                        int value = Integer.valueOf(binding.tvItemListStreamViewers.getText());
                        int randomValue = +view[new Random().nextInt(view.length - 1)];
                        if (value + randomValue >= 0)
                            binding.tvItemListStreamViewers.setText(String.valueOf(value + randomValue));
                        handler.postDelayed(this, duration[new Random().nextInt(duration.length - 1)]);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }

                }
            });
        }

        private void bindTo() {
        }

        private void clear() {
            binding.ivItemListStream.setImageResource(R.drawable.streams_placeholder);
            binding.tvItemListStreamTitle.setText("");
            binding.tvItemListStreamSummary.setText("");
        }
    }


}

File: app/src/main/java/n7/ad2/ui/streams/StreamsPagedListAdapter.kt
package n7.ad2.ui.streams

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import n7.ad2.R
import n7.ad2.databinding.ItemListStreamBinding
import n7.ad2.ui.streams.domain.vo.VOSimpleStream
import n7.ad2.ui.streams.domain.vo.VOStream
import n7.ad2.utils.ImageLoader

class StreamsPagedListAdapter(
    private val imageLoader: ImageLoader,
    private val onStreamClick: (vOSimpleStream: VOSimpleStream) -> Unit,
) : PagingDataAdapter<VOStream, RecyclerView.ViewHolder>(DiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = when (viewType) {
        R.layout.item_list_stream -> StreamViewHolder.from(parent, imageLoader)
        else -> throw UnsupportedOperationException("could not find ViewHolder for $viewType")
    }

    override fun getItemViewType(position: Int) = when (getItem(position)) {
        is VOSimpleStream -> R.layout.item_list_stream
        else -> throw UnsupportedOperationException("could not get type for $position")
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = getItem(position)
        when (holder) {
            is StreamViewHolder -> holder.bindTo(item as VOSimpleStream, onStreamClick)
        }
    }

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        when (holder) {
            is StreamViewHolder -> holder.unbind()
        }
    }

    class StreamViewHolder(
        private val binding: ItemListStreamBinding,
        private val imageLoader: ImageLoader,
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bindTo(
            vOSimpleStream: VOSimpleStream,
            onStreamClick: (vOSimpleStream: VOSimpleStream) -> Unit,
        ) = binding.apply {
            binding.tvTitle.text = vOSimpleStream.title
            binding.root.setOnClickListener { onStreamClick(vOSimpleStream) }
            imageLoader.load(binding.ivStreamImage, vOSimpleStream.imageUrl, R.drawable.streams_placeholder)
        }

        fun unbind() {
            binding.root.setOnClickListener(null)
            imageLoader.clear(binding.ivStreamImage)
        }

        companion object {
            fun from(
                parent: ViewGroup,
                imageLoader: ImageLoader,
            ): StreamViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ItemListStreamBinding.inflate(layoutInflater, parent, false)
                return StreamViewHolder(binding, imageLoader)
            }
        }

    }

    private class DiffCallback : DiffUtil.ItemCallback<VOStream>() {
        override fun areItemsTheSame(oldItem: VOStream, newItem: VOStream) = oldItem::class == newItem::class
        override fun areContentsTheSame(oldItem: VOStream, newItem: VOStream) = oldItem == newItem
    }
}